app/models/request/statemachine.rb

```diff
@@ -192,7 +192,13 @@ def on_blocked; end
   def on_hold; end
 
   def failed_upstream!
-    fail_from_upstream! unless failed?
+    # Don't transition it again if it's already failed
+    return if failed?
+
+    # Only transition it if *all* upstream requests are failed, not just the one we came from.
+    return unless upstream_requests.all?(&:failed?)
```

Comment:
Similarly here, it probably also makes sense to include cancelled? There's a `terminated?` check that covers both states already.